### PR TITLE
Add shutdown timeout

### PR DESCRIPTION
**Pull request by urso**
Introduce signalWait structure to filebeat waiting for stop signal after
fb.done was closed. If shutdown_timeout is configured, filebeat will either
stop after timeout or after all enqueued events have been successfully
published and written by registrar.

I hope the signalWait idea can be reused for implementing run-once mode as well.

### CODE REVIEW

File: filebeat/beater/signalwait.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+func (s *signalWait) AddTimer(t *time.Timer) {
+	s.Add(func() { <-t.C })
+}
+
+func (s *signalWait) AddTimeout(d time.Duration) {
+	s.AddTimer(time.NewTimer(d))
+}
+
+func (s *signalWait) Signal() {
```
Comment (ruflin) [inline filebeat/beater/signalwait.go]: What is this needed for?

Comment (urso) [inline filebeat/beater/signalwait.go]: It's for sending inserting a signal. E.g. having received Ctrl-C one can use `Signal()` to insert a signal into the waiter that triggers right now, such that `Wait` will not block.

e.g.
```
<- fb.done
if whatever {
    sigWait.Signal()
}
```

Comment (ruflin) [inline filebeat/beater/signalwait.go]: Got it

**Comment by ruflin** [comment]
LGTM

File: filebeat/beater/filebeat.go [hunk 2]
```diff
@@ -106,6 +125,12 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	// Blocks progressing. As soon as channel is closed, all defer statements come into play
 	<-fb.done
 
+	if fb.config.ShutdownTimeout > 0 {
+		// Wait for either timeout or all events having been ACKed by outputs.
```
Comment (ruflin) [inline filebeat/beater/filebeat.go]: We should add here a log message that shutdown will wait for time X to shutdown or success

**Pull request merged**